spec.md

```diff
@@ -2709,6 +2713,67 @@ need only respect the 2-state behavior of a circuit.  This is a limitation on
 the scope of what behavior is observable (i.e., a relaxation of the
 ["as-if"](https://en.wikipedia.org/wiki/As-if_rule) rule).
 
+## Indeterminate Values
+
+An indeterminate value represents a value which is unknown or unspecified.  
+Indeterminate values are generally implementation defined, with constraints 
+specified below.  An indeterminate value may be assumed to be any specific 
+value, at an implementation's descresion, if, in doing so, all observable 
+behavior is as if the indeterminate value always took the specific value.
+
+This allows transformations such as the following, where when `a` has an 
+indeterminate value, the implementation chooses to consistently give it a value 
+of 4.  There is no visibility of a when it has an indeterminate value which does 
+not see it as 4.
+``` firrtl
+module IValue :
+  output o : UInt<8>
+  input c : UInt<1>
+
+  wire a : UInt<8>
+  a is invalid
+  when c :
+    a <= UInt<3>("h4")
+  o <= a
+```
+is transfromed to:
+``` firrtl
+module IValue :
+  output o : UInt<8>
+  input c : UInt<1>
+
+  o <= UInt<3>("h4")
+```
+Note that it is equally correct to produce:
+``` firrtl
+module IValue :
+  output o : UInt<8>
+  input c : UInt<1>
+
+  wire a : UInt<8>
+  when c :
+    a <= UInt<3>("h4")
+   else :
+     a <= UInt<3>("h42")
+  o <= a
+```
+
+The behavior of constructs which cause indeterminate values is implementation 
+defined with the following constraints.  
+* Registers initialization is of a uniform code pattern.  If code is generated 
+to randomly initialize some registers (or 0 fill them, etc), it should be 
+generated for all registers.
+* All observations of an expression with indeterminate value must see the same 
+value at runtime.  Multiple readers of a value will see the same runtime value.
```

Comment:
Does the "all observations" need to hold true across transforms that may create a copy or replace the construct/wire/reg?  As written I would tentatively say it does since this is the source-level semantics and must be true regardless of how the construct gets lowered.  Primarily thinking of inlining as where this would come up, although elsewhere too.  Similarly, I think this may cause different behavior  in a circuit before/after dedup if different values were used in each or possibly before/after lowering an aggregate register (to registers for each ground type).

Mostly lets not cause ourselves the need to take measures like assign these values early and track them throughout + across optimizations.

Standardizing on '0' does avoid this issue :upside_down_face: .

Reply:
The intent is certainly observations of runtime instances not input syntactic instances.
Maybe saying observations of unique runtime instances would help.

Reply:
Would it help to lean on a new IR operation, like `freeze`, that would help clarify at what point these guarantees kick in as opposed to the somewhat more amorphous "expression with indeterminate value".

E.g.:

```
wire a: UInt<1>
a is invalid

node b = freeze(a)
node c = freeze(a)

assert(b == b) ; this is always true
assert(b == c) ; this is always false?
```

Reply:
Shouldn't the second sentence be:
> Multiple readers of _the same expression_ will see the same runtime value.
?

I am also wondering if this would benefit from a clarification about the temporal aspect. I.e., the:
> Multiple readers of the same expression, _in the same time step_ will see the same runtime value.